### ISSUE TITLE
How to style list's "remove", "up" and "down" buttons?

### ISSUE THREAD

**Issue by Xananax** [gcanti/tcomb-form#254]
I'm having a hard time wrapping my head around the best way to customize the those buttons.

The way I see it, currently, if I wanted to replace them with icons (respectively `✖`, `▲` and `▼`), the process is quite involved.

I tried following up the codepath where the buttons are created (for bootstrap templates) and it seems to me the following snippets are the important ones:

In `components.js`, in the `getItems()` method of the `List` component, the buttons options get created
```js
const buttons = [];
      if (!options.disableRemove) { buttons.push({ label: i18n.remove, click: this.removeItem.bind(this, i) }); }
      if (!options.disableOrder)  { buttons.push({ label: i18n.up, click: this.moveUpItem.bind(this, i) }); }
      if (!options.disableOrder)  { buttons.push({ label: i18n.down, click: this.moveDownItem.bind(this, i) }); }
```

In `templates/bootstrap.js`, in the `list` template, the buttons get created:
```js
bootstrap.getCol({
          breakpoints: {sm: 4, xs: 6},
          children: bootstrap.getButtonGroup(item.buttons.map(function (button, i) {
            return bootstrap.getButton({
              click: button.click,
              key: i,
              label: button.label
            });
          }))
        })
```

Then in the `getButton.js` file of the `uvdom-bootstrap` module, the button actually gets instantiated.

So, in order to customize the template, one would have to:

1. reimplement (or extend) the `List` Component
2. reimplement the `list` template
3. as to not stray too far away from the core code (in order to reproduce changes easily), one has to import and learn to use `uvdom` and `uvdom-boostrap` (arguably not very hard, but still a slight additional learning curve).

It's a complex process because both the List component and the list template make use of a bunch of helper functions which are not always exposed. They are also pretty complex (in other words, easy to break when the main library changes something).

Alternatively, one could create a factory that reimplements `onItemChange`, `removeItem`, `moveUpItem`, and `moveDownItem`, but you'd also have to reimplement `validate`, errors handling, and so on (so basically actually reimplementing the `List` component almost in it's entirety).

It seems to me that it would be better if the buttons were their own components, so they could be skinned just like everything else:

```js
f.form.Form.templates.upButton = template
```
or:

```js
options = {
    templates:{
         upButton:template
    }
}
```

Am I missing something? Is there an easier way to accomplish this that I'm not seeing?

**Comment by Xananax** [comment]
By the way, I realize I can get away with 

```js
t.form.Form.i18n = {
	optional: ''
,	required: ''
,	add: '➕'
,	remove: '✖'
,	up: '▲'
,	down: '▼'
}
```

But that's pretty limited in terms of making those buttons consistent with the rest of my theme.

**Comment by minedeljkovic** [comment]
What makes it really hard to do it in template only is that there is no way to properly differentiate between remove, up, down buttons (well you can hack with labels).

Maybe some improvement in List options would make it easier (maybe even enable styling without template):
```javascript
var options = {
	remove: {
		className: 'someRemoveButtonClass'
	},
	up: {
		template: someUpButtonTemplate
	}
}
```

or maybe to make it more inline with other simple components options:
```javascript
var options = {
	remove: {
		attrs: {
			className: 'someRemoveButtonClass'
		},
		disabled: true
	},
	up: {
		template: someUpButtonTemplate
	}
}
```

**Comment by Xananax** [comment]
A simple fix for styling would be to at least give the buttons default classes

```js
const buttons = [];
      if (!options.disableRemove) { buttons.push({ label: i18n.remove, click: this.removeItem.bind(this, i),classname:'btn-remove'}); }
      if (!options.disableOrder)  { buttons.push({ label: i18n.up, click: this.moveUpItem.bind(this, i),className:'btn-move-up' }); }
      if (!options.disableOrder)  { buttons.push({ label: i18n.down, click: this.moveDownItem.bind(this, i),className:'btn-move-down' }); }
```

This would give some room to play around.

I realize this should not be the concern of the factory, and should be handled in the template, and I agree with the choice in principle, but in practice, it might be worth it to favour ease of use over purity. Currently, from the template, there is no way to know which button does what (other than by checking the label, which is a very brittle way). Maybe a custom `btnType` property that the template could use?

In the factory:
```js
if (!options.disableRemove) { buttons.push({ label, click,btnType:'remove'}); }
```

in the bootstrap template:
```js
return bootstrap.getButton({
    click: button.click,
    key: i,
    label: button.label,
   className:(button.btnType?'btn-'+button.btnType:'')
});
```

(not sure uvdom uses `className`, this is just pseudo-code).

This doesn't seem like a big change in architecture and would open the door for further customization. I'd be willing to submit a PR, but I'll wait for @gcanti to pitch in before making a move.

Maybe it would be even better if the template returned something like this:
```js
<a role="button" className={`btn btn-${btnType}`}>
    <span className={`btn-label btn-label-${btnType}`}>{label}</span>
</a>
```
Having an `<a/>` and a nested `<span/>` would allow for all but the most complex stylings.

**Comment by gcanti** [comment]
Hi, sorry for the delay. 

Thanks a lot for your suggestions, very interesting.

> So, in order to customize the template, one would have to...

@Xananax currently no need for 1. and 3. just set the `template` option

```js
function myListTemplate(locals) {
  // you can use jsx here
  return <div>myListTemplate...</div>;
}

const Type = t.struct({
  tags: t.list(t.String)
});

const options = {
  fields: {
    tags: {
      template: myListTemplate // <= local override
    }
  }
};
```

> What makes it really hard to do it in template only is that there is no way to properly differentiate between remove, up, down buttons

> A simple fix for styling would be to at least give the buttons default classes

@minedeljkovic @Xananax I agree, that would give some room to basic styling, will add a `type` property to the button locals https://github.com/gcanti/tcomb-form/blob/master/src/components.js#L725

then the template could add proper style classes https://github.com/gcanti/tcomb-form/blob/master/src/templates/bootstrap/list.js#L68

> It seems to me that it would be better if the buttons were their own components, so they could be skinned just like everything else:

Seems a great idea, what about something like this? (renderers factored out in the list template)

https://github.com/gcanti/tcomb-form/blob/master/src/templates/bootstrap/list.js

and then override like this

```js
// override the row buttons...
t.form.Form.templates.list.renderRowButton = function (button) {
  return <button onClick={button.click}>{button.label}</button>;
};

// ...or override the add buttons, and so on...
t.form.Form.templates.list.renderAddButton = function (button) {
  return <button onClick={button.click}>{button.label}</button>;
};

const Type = t.struct({
  tags: t.list(t.String)
});
```

seems quite easy, what do you think?

**Comment by minedeljkovic** [comment]
I think the templates flexibility that you added are step in the right direction.
It reminds me a lot of a concept (from godforsaken platform that I spent a lot time working on) of Template parts (http://www.jeff.wilcox.name/2010/04/template-part-tips/).

Increase in flexibility is obvious. You can change parts on an existing template globally or locally (just clone global template, change that little part that must be different in just that one place and set that new template through options).

Of course (as with everything) it takes skill to define "just enough" parts, not to make everything over complex, but the tools are here.

For me, nicely done!

**Comment by gcanti** [comment]
Just added a `clone` method to templates, now you can do local overrides:

**Global override**

```js
function renderAddButton(locals) {
  return <button onClick={locals.add.click}>{locals.add.label}</button>;
}

// every list is affected
t.form.Form.templates.list.renderAddButton = renderAddButton;

const Type = t.struct({
  a: t.list(t.String),
  b: t.list(t.String)
});
```

**Local override**

```js
function renderAddButton(locals) {
  return <button onClick={locals.add.click}>{locals.add.label}</button>;
}

const myList = t.form.Form.templates.list.clone();
myList.renderAddButton = renderAddButton;

const Type = t.struct({
  a: t.list(t.String),
  b: t.list(t.String)
});

const options = {
  fields: {
    b: {
      templates: {
        list: myList // <= only list `b` is affected, list `a` use defaults
      }
    }
  }
};
```

**Comment by minedeljkovic** [comment]
When I was writing clone, I meant more of a standard shallow pattern like `Object.assign()`. Do you find that deep clone is necessary @gcanti?
If so, current version can be a bit misleading. In previous example if `myList.clone()` is called it seeems that clone of `t.form.Form.templates.list` will be received.

**Comment by gcanti** [comment]
> Do you find that deep clone is necessary @gcanti?

The problem I see with the actual implementation (i.e. attaching the `render*` functions to the function `list`) is that `list` will retain the references to the `render*` functions captured during its definition (and so will do the `render*` functions as well), thus  if `list` is not cloned `list.renderSomething = myRenderSomething` doesn't work (is ignored if my tests are correct).
Do you have suggestions for an alternative implementation?

**Comment by minedeljkovic** [comment]
Yes, I was completely off with my view of clone, forgetting that templates are of course functions. I was quick to fix my comment, but you were quicker. :-)
There are of course some solutions for function cloning (http://stackoverflow.com/a/11230005), but it is not very elegant (which can be a sign of need for a different approach). I will be thinking on this.

**Comment by minedeljkovic** [comment]
Slightly modified version maybe?

list component:
```js
function create(overrides = {}) {

  function list(locals) {
    ...
  }
  
  list.extend = create;
  
  list.renderAddButton = overrides.renderAddButton || function (locals) {
    // default implementation
  };

  return list;
}

export default create();
```

then local override:
```js
function renderAddButton(locals) {
  return <button onClick={locals.add.click}>{locals.add.label}</button>;
}

const myList = t.form.Form.templates.list.extend({ renderAddButton });
myList.renderAddButton = renderAddButton;
```

**Comment by gcanti** [comment]
@minedeljkovic great idea, one last problem though

```js
function myRenderAddButton(locals) {
  return <button onClick={locals.add.click}>{locals.add.label}</button>;
}

console.log(t.form.Form.templates.list.renderAddButton.name); // => renderAddButton (default) OK!
const myList1 = t.form.Form.templates.list.extend({renderAddButton: myRenderAddButton});
console.log(myList1.renderAddButton.name); // => myRenderAddButton OK!
const myList2 = myList1.extend();
console.log(myList2.renderAddButton.name); // => renderAddButton OUCH!
```

modifying `list.extend`

```js
function create(overrides = {}) {

  function list(locals) {
    ...
  }

  list.renderAddButton = overrides.renderAddButton || function (locals) {
    // default implementation
  };

  list.extend = function (overrides = {}) {
    return create({...list, ...overrides}); // keep previous overrides
  };

  return list;
}

export default create();
```

As a (beneficial?) side effect, you could even modify the overrides after the `extend` call

```js
function myRenderAddButton(locals) {
  return <button onClick={locals.add.click}>{locals.add.label}</button>;
}

const myList1 = t.form.Form.templates.list.extend();
myList1.renderAddButton = myRenderAddButton; // this now works well with extend()

const myList2 = t.form.Form.templates.list.extend({
  renderAddButton: myRenderAddButton
});

console.log('default renderer', t.form.Form.templates.list.renderAddButton.name); // => renderAddButton (default) OK!

console.log('myList1', myList1.renderAddButton.name); // => myRenderAddButton OK!
console.log('myList1.extend()', myList1.extend().renderAddButton.name); // => myRenderAddButton OK!

console.log('myList2', myList2.renderAddButton.name); // => myRenderAddButton OK!
console.log('myList2.extend()', myList2.extend().renderAddButton.name); // => myRenderAddButton OK!
```

One minor pick: not sure about the name, `extend` seems too OOPish... I'd keep `clone`

**Comment by minedeljkovic** [comment]
Yes, @gcanti great fix!

> One minor pick: not sure about the name, extend seems too OOPish... I'd keep clone

I don't have preference towards extend (it even bothers me too), so of course. 
I am even bothered with `overrides`. (I was thinking of calling it `template`, but a ui template's function template, ugh... `config` is of course vague..)

**Comment by gcanti** [comment]
> As a (beneficial?) side effect...

Mhhh it's tricky... after additional tests there are some failing corner cases. I'd end up with this implementation

```js
function create(overrides = {}) {

  ...

  list.clone = function clone(newOverrides = {}) {
    return create({...overrides, ...newOverrides});
  };

  return list;
}
```

and discourage mutability

```js
// DON'T DO THIS
const myList1 = t.form.Form.templates.list.clone();
myList1.renderAddButton = myRenderAddButton;

// do this instead
const myList1 = t.form.Form.templates.list.clone({
  renderAddButton: myRenderAddButton
});
```

moreover having only one way to get new templates seems a better practice.

**Comment by gcanti** [comment]
Released in [v0.7.7](https://github.com/gcanti/tcomb-form/releases/tag/v0.7.7)
